idaes/gas_solid_contactors/properties/oxygen_iron_OC_oxidation/hetero_reactions.py

```diff
@@ -0,0 +1,517 @@
+                               'length': 'm',
+                               'mass': 'kg',
+                               'amount': 'mol',
+                               'temperature': 'K',
+                               'energy': 'kJ'})
+
+
+class _ReactionBlock(ReactionBlockBase):
+    """
+    This Class contains methods which should be applied to Reaction Blocks as a
+    whole, rather than individual elements of indexed Reaction Blocks.
+    """
+    def initialize(blk, outlvl=idaeslog.NOTSET,
+                   optarg={'tol': 1e-8}, solver='ipopt'):
+        '''
+        Initialisation routine for reaction package.
+
+        Keyword Arguments:
+            outlvl : sets output level of initialization routine
+                 * 0 = Use default idaes.init logger setting
+                 * 1 = Maximum output
+                 * 2 = Include solver output
+                 * 3 = Return solver state for each step in subroutines
+                 * 4 = Return solver state for each step in routine
+                 * 5 = Final initialization status and exceptions
+                 * 6 = No output
+            optarg : solver options dictionary object (default=None)
+            solver : str indicating whcih solver to use during
+                     initialization (default = "ipopt")
+        Returns:
+            None
+        '''
+        init_log = idaeslog.getInitLogger(blk.name, outlvl, tag="reactions")
+        solve_log = idaeslog.getSolveLogger(blk.name, outlvl, tag="reactions")
+
+        init_log.info_high('Starting initialization')
+
+        # TODO - Update in the future as needed
+        # Get a single representative block for getting config arguments
+        for k in blk.keys():
+            break
+
+        # Fix state variables if not already fixed
+        # Fix state variables of the primary (solid) state block
+        state_var_flags = fix_state_vars(blk[k].config.solid_state_block)
+
+        # Fix values of secondary (gas) state block variables if not fixed,
+        # as well as the solid density variable.
+        # This is done to keep the initialization problem square
+        Cflag = {}  # Gas concentration flag
+        Dflag = {}  # Solid density flag
+
+        for k in blk.keys():
+            for j in blk[k]._params.gas_component_list:
+                if blk[k].gas_state_ref.dens_mol_comp[j].fixed is True:
+                    Cflag[k, j] = True
+                else:
+                    Cflag[k, j] = False
+                    blk[k].gas_state_ref.dens_mol_comp[j].fix(
+                            blk[k].gas_state_ref.dens_mol_comp[j].value)
+            if blk[k].solid_state_ref.dens_mass_skeletal.fixed is True:
+                Dflag[k] = True
+            else:
+                Dflag[k] = False
+                blk[k].solid_state_ref.dens_mass_skeletal.fix(
+                        blk[k].solid_state_ref.dens_mass_skeletal.value)
+
+        # Set solver options
+        opt = SolverFactory(solver)
```

Comment:
Minor comment, but to save time you could move the creation of the solver object to within the `if` statement that runs the solver. This way, we only build the solver object if it is actually needed.

Reply:
Also look at the new `get_solver` method.